woudc_data_registry/models.py

```diff
@@ -1163,12 +1250,113 @@ def sync(ctx):
         Station,
         Instrument,
         Deployment,
-        DataRecord
+        DataRecord,
     ]
+    contribution_models = list()
+
+    """function to form all the contributions from the other models lists"""
+    def build_contributions(contribution_models, project_models,
+                            station_models, instrument_models, country_models,
+                            deployment_models, contributor_models):
+
+        """turn lists holding essential information into dictionaries with key
+        that is common with another list in order to allow for fast lookup"""
+        country_dict = {country_models[i].country_id: country_models[i]
+                        for i in range(0, len(country_models))}
+        instrument_dict = {instrument_models[i].name: instrument_models[i]
+                           for i in range(0, len(instrument_models))}
+        contributor_dict = {contributor_models[i].contributor_id:
+                            contributor_models[i]
+                            for i in range(0, len(contributor_models))}
+
+        """deployment_dict will have stations as
+        keys in order to help connect them to projects """
+        deployment_dict = {deployment_models[i].station_id:
+                           deployment_models[i]
+                           for i in range(0, len(deployment_models))}
+
+        """contribution dict used to check for duplicate contribution id"""
+        contribution_dict = {contribution_models[i].contribution_id:
+                             contribution_models[i]
+                             for i in range(0, len(contribution_models))}
+
+        for key in instrument_dict:
+
+            for station in station_models:
+
+                """station info"""
+                station_id = station.station_id
+                station_name = station.station_name_id
```

Comment:
Really glad you got the right idea with the dictionaries, and it will definitely be handy for some parts of this.

But for a lot of these you can use sqlalchemy's relationships/joins to do things like `station_id = instrument.station_id` or `country_name_en = instrument.station.country.name_en` and stuff like that, so that the instrument model on its own can get you most of the information you need.